csv parser used in import alias refactor

SytzeAndr opened the pull request:
<!-- Anything that looks like this is a comment and can't be seen after the Pull Request is created. -->

# PR Summary
The old implementation used a for loop with lots of if statements to parse a csv string, which is used in the `import-alias` cmdlet. I think the old implementation is bug sensitive and hard to read. A new implementation can have a reduced complexity and uses the `StringReader` and `StringBuilder` to convert a string in csv format to a collection of elements, while keeping functionality intact. Experimental analysis shows this new implementation performs better, especially for large Strings.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: <!-- Number/link of that issue here -->
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
        - [ ] [Add `[feature]` to your commit messages if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)

SytzeAndr committed:
regex expression to read csv string and ignore quotes

SytzeAndr committed:
added reference

SytzeAndr committed:
cleanup and delimiter set included

msftclas commented:
[![CLA assistant check](https://cla.opensource.microsoft.com/pull/badge/not_signed)](https://cla.opensource.microsoft.com/PowerShell/PowerShell?pullRequest=9091) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.opensource.microsoft.com/PowerShell/PowerShell?pullRequest=9091) before we can accept your contribution.<br/><br/>:x: SytzeAndr [sign now](https://cla.opensource.microsoft.com/PowerShell/PowerShell?pullRequest=9091)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.microsoft.com/check/PowerShell/PowerShell?pullRequest=9091) it.</sub>

SytzeAndr committed:
code comment issues from CI fixed

SytzeAndr committed:
removed comment that gave CI issues, since it is not really necessary

vexx32 commented:
@SytzeAndr the `#` case is for when the CSV contains a type hint like it used to in Windows PowerShell -- CSVs generated without the `-NoTypeInformation` would have a one-line header with something like this:
```
PS Z:\> [pscustomobject]@{a = 1} | ConvertTo-Csv
#TYPE System.Management.Automation.PSCustomObject
"a"
"1"
```

SytzeAndr committed:
removal of toArray and wrong String usage

SytzeAndr committed:
returns empty collection when conditions met

reuse of the old implementation

SytzeAndr commented:
> @SytzeAndr the `#` case is for when the CSV contains a type hint like it used to in Windows PowerShell -- CSVs generated without the `-NoTypeInformation` would have a one-line header with something like this:
> 
> ```
> PS Z:\> [pscustomobject]@{a = 1} | ConvertTo-Csv
> #TYPE System.Management.Automation.PSCustomObject
> "a"
> "1"
> ```

Thanks, I'll then reuse the part from the old implementation for when to return an empty collection.

iSazonov commented:
@SytzeAndr Thanks for your contribution!

We are working on enhancing the cmdlet and have plans to improve performance.
I am concerned that your change affects performance and adds extra allocations. Did you measure?

SytzeAndr commented:
@iSazonov I haven't tested it yet, but performance is one of the reasons I think refactoring this code would make sense. The old implementation is synchronous, but it should be possible to make this an asynchronous process instead of iterating over all the characters which should be faster. 

I haven't set up the testing environment locally (so therefore WIP), but when I do and the tests are passing I'll do a measure in terms of performance against the old implementation aswell.

powercode commented:
Are you aware that this method is only called from the `Import-Alias` cmdlet?

The implementation of `Import-Csv` is in the file CSVCommands.cs, and the bulk of the work is done in the ImportCsvHelper class. That said, almost all of the cost of this operation is in creating the resulting objects. IIRC, the stream parsing was below 5%. And while the first rule of performance is to measure, in this case I would venture to speak from experience and say that using Regex is not the way to improve the performance.

powercode commented:
@SytzeAndr If you want to improve performance, I would suggest a different approach. Start with a scenario you know is slow. Then run it under a profiler, and try to understand if there are any bottlenecks.

Then try to understand the code around the bottleneck - think about it, why is it written the way it is, how could it be done instead etc. Keep recordings of the original scenario. Try out changes, and measure them, while at the same time making sure that the tests that covers your changes still passes.

The area you have selected here is very hard to improve significantly upon, especially since it is not in the hot path.

SytzeAndr committed:
implementation with reader

SytzeAndr committed:
implementation with reader

removed old commented code and implemented a version using the System.IO.Stringreader

SytzeAndr committed:
codefactor warnings

SytzeAndr committed:
more codefactor suggestions

SytzeAndr committed:
more codefactor suggestions regarding wrongly placed brackets and comments

SytzeAndr committed:
more codefactor regarding brackets

SytzeAndr committed:
remove commented code

SytzeAndr committed:
added a stringbuilder

SytzeAndr committed:
rename for clarity

SytzeAndr committed:
consistency with brackets

SytzeAndr commented:
some updates on this PR:

I've implemented the method by using the StringReader and StringBuilder. It currently passes all CI tests, so I assume my implementation is fine.

I assume use of the [Stringbuilder](
https://docs.microsoft.com/en-us/dotnet/api/system.text.stringbuilder?view=netcore-2.2) has a positive impact on performance since the `string` class builds a new string at each operation while the stringbuilder maintains it. 

I've reduced the complexity of this method according to CodeFactor (and is regarded as being a fixed issue to the repo). Although I still think it can be improved by splitting some things and might do so.

@powercode I used this PR aswell to get a bit used to contributing to powershell, and I chose to tackle an issue which could be fixed with some refactoring. Therefore, I didn't start with analyzing the bottlenecks since I had to set up things yet and wanted to start small by just refactoring some code. I agree with you in the sense that it would make more sense to analyze from a bit higher perspective to find the true bottlenecks. For a next PR, I'll try to do so, but a main reason for this PR is to just have something improved in powershell's code base, have my test environment set up, see what contributing to powershell is like, those sort of things. 

So the next step for me is to test it for performance against the old implementation, after that I'll remove the WIP tag (assuming performance is not worse). 
Is there some simple way of checking the performance/runtime by checking the CI runtime results for different branches, or is it better to do some local performance-based-tests myself of the different implementations and report them here?

Geweldig commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/Csv.cs:
```diff
@@ -20,92 +22,69 @@ internal CSVHelper(char delimiter)
-                if (c == Delimiter)
+                char nextChar = (char)reader.Read();
+
+                // if next character was delimiter, add string to builder and clear builder
```
This comment does not seem to reflect what the code does, I assume you meant `add string to result and clear builder`.

Geweldig commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/Csv.cs:
```diff
@@ -20,92 +22,69 @@ internal CSVHelper(char delimiter)
-            bool inQuote = false;
-            for (int i = 0; i < csv.Length; i++)
+            var reader = new StringReader(csv);
+            StringBuilder stringBuilder = new StringBuilder();
```
I would suggest renaming `stringBuilder`  to something which reflects what is stored in the builder, instead of reflecting what type of object it is. Perhaps something such as `wordBuffer` as you're building up a word and clear it when the word has been read fully?

SytzeAndr commented:
@Geweldig makes sense, thanks for the suggestions!

SytzeAndr committed:
Implemented changes @Geweldig suggested

SytzeAndr commented:
I did some test and the performance seems to be improved, especially for large strings. I took the largest csv file from [here](https://sample-videos.com/download-sample-csv.php) called SampleCSVFile_53000kb.csv which contained 500.000 rows and is 53kb large. Then, I created some other very small csv file `very_small.csv` with the content `this, is, a, "very", small, file`. I did an experiment 100.000 times for the small file, and 100 times for the large file. The log is below.

For the large file, the performance seems to be twice as good (70s to 35s). For the small file, the performance is better, but not significantly (11.1s to 10.8s).

I'll remove the WIP tag and have this open for review.

```
file:csv_files/SampleCSVFile_53000kb.csv
iterations:100
useold:True
time elapsed:74262ms

file:csv_files/SampleCSVFile_53000kb.csv
iterations:100
version:False
time elapsed:35953ms

file:csv_files/very_small.csv
iterations:100000
useold:True
time elapsed:11080ms

file:csv_files/very_small.csv
iterations:100000
version:False
time elapsed:10759ms
````

iSazonov commented:
@SytzeAndr Can you share the test script?

SytzeAndr commented:
I used the following script. `CSVHelper` corresponds to my implementation, for `Csv_old` I copied code of the old implementation.

```
using System;
using Microsoft.PowerShell.Commands;
using System.IO;
using System.Diagnostics;

namespace csv_helper_new
{
    class Program
    {
        static void Main(string[] args)
        {
            bool useold = true;
            //args are files to read
            int iterations = 100;
            Stopwatch stopwatch = new Stopwatch();
            foreach (String a in args) {
                stopwatch.Reset();
                Console.WriteLine("file:"+a+"\n"+"iterations:"+iterations+"\nuseold:"+useold);
                stopwatch.Start();
                for (int i=0;i<iterations;i++) {
                    if (useold) {
                        Csv_old cSVHelper = new Csv_old(',');
                        cSVHelper.ParseCsv(File.ReadAllText(a)).ToString();
                    } else {
                        CSVHelper cSVHelper = new CSVHelper(',');
                        cSVHelper.ParseCsv(File.ReadAllText(a)).ToString();
                    }
                }
                stopwatch.Stop();
                Console.WriteLine("time elapsed:"+stopwatch.ElapsedTicks/1000+"ms");
            }            
        }
    }
}
```

iSazonov commented:
@SytzeAndr Please clarify what do you want make faster? The `ParseCsv()` method is only used in Import-Alias cmdlet - do you really want to improve the cmdlet?

LMiljak commented:
I think the main goal of this PR is to have the original code refactored (as the complexity of the method was quite high), without sacrificing performance. Making it faster was not necessarily the goal.

SytzeAndr commented:
@iSazonov I wanted to improve code quality since I found the method hard to read without affecting performance in a negative way. Since you commented

> I am concerned that your change affects performance and adds extra allocations. Did you measure?

I decided to do experimental analysis, for both a very large and small string, to show that the new implementation is not worse, and concluded it performs better. A bad performance of the `import-alias` cmdlet is not the reason for the PR, but a bad performance should not occur due to some PR.

iSazonov commented:
I respect your efforts. Nevertheless, they are too wasteful. This code is very rarely used and only for importing a few aliases. Therefore, your tests for large files are useless.

I will continue to review the code without these performance considerations. Please update the PR description.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/Csv.cs:
```diff
@@ -20,92 +22,69 @@ internal CSVHelper(char delimiter)
         /// </summary>
         internal char Delimiter { get; } = ',';
 
+        internal char Quote { get; } = '"';
```
I don't see a value for this. Please use const `'"'` in code - it is more readable.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/Csv.cs:
```diff
@@ -20,92 +22,69 @@ internal CSVHelper(char delimiter)
         internal Collection<string> ParseCsv(string csv)
-        {
+        {        
+            csv = csv.Trim(); 
```
We could consider using ReadOnlySpan<char>. It is optional.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/Csv.cs:
```diff
@@ -20,92 +22,69 @@ internal CSVHelper(char delimiter)
+            if (wordBuffer.ToString() != string.Empty) 
             {
-                result.Add(tempString);
+                result.Add(wordBuffer.ToString());
```
We evaluate `wordBuffer.ToString()` twice - please avoid.

SytzeAndr committed:
iSazonov suggestions implemented

SytzeAndr committed:
extra parentheses

SytzeAndr committed:
old way, adding last string outside of the while loop since the build was breaking

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/Csv.cs:
```diff
@@ -27,85 +29,59 @@ internal CSVHelper(char delimiter)
+                else if (nextChar == '"') 
                 {
-                    switch (c)
+                    bool isinQuotes = true;
```
Please replace isinQuotes with isInQuotes

SytzeAndr replied:
Unfortunately this raises a codeFactor error regarding Hungarian notation. I changed it to `betweenQuotes`

![image](https://user-images.githubusercontent.com/9801575/54298615-60fd7380-45b9-11e9-9e6b-a9cf76a7a8ba.png)

SytzeAndr committed:
isinQuotes -> isInQuotes

SytzeAndr committed:
isInQuotes -> betweenQuotes

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/Csv.cs:
```diff
@@ -27,85 +29,59 @@ internal CSVHelper(char delimiter)
-            {
-                result.Add(tempString);
+            string lastWord = wordBuffer.ToString();
+            if (lastWord != string.Empty) {
```
```suggestion
            if (lastWord != string.Empty)
            {
```

iSazonov committed:
Update src/Microsoft.PowerShell.Commands.Utility/commands/utility/Csv.cs

Co-Authored-By: SytzeAndr <s.p.e.andringa@student.tudelft.nl>

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/Csv.cs:
```diff
@@ -27,85 +29,59 @@ internal CSVHelper(char delimiter)
+                else if (nextChar == '"') 
                 {
-                    switch (c)
+                    bool betweenQuotes = true;
```
We already use `inQuotes` in tokenizer so maybe it is just better.

SytzeAndr replied:
At the moment I'm not sure which you currently think is best

vexx32 replied:
Hungarian notation warning can be ignored for now. 

Dongbo has a PR pending that is disabling some of the CodeFactor rules the team don't agree with. 🙂

SytzeAndr replied:
then `inQuotes` it is!

SytzeAndr committed:
betweenQuotes -> inQuotes

SytzeAndr committed:
Merge branch 'csv_refactor' of https://github.com/Geweldig/PowerShell into csv_refactor

powercode commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/Csv.cs:
```diff
@@ -27,85 +29,60 @@ internal CSVHelper(char delimiter)
+                char nextChar = (char)reader.Read();
+
+                // if next character was delimiter or we are at the end, add string to result and clear builder
+                // else if next character was quote, perform reading untill next quote and add it to builder
```
untill -> until

daxian-dbw commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/Csv.cs:
```diff
@@ -27,85 +29,60 @@ internal CSVHelper(char delimiter)
+                        
+                        if (nextChar == '"') 
+                        {
+                            inQuotes = false;
```
You are not handling the input like this: `"abc""def"`. This will happen when the original value contains `"` in it.

SytzeAndr replied:
Sorry for my late reply. 

I'm not entirely sure what you mean. As far as I can see the functionality of the old and new implementation correspond, in the sense that the `"` is excluded in the content. `"abc""def"` should parse to `[abc,def]`. Did you spot an inconsistency with respect to the old implementation or are you pointing at something else?

daxian-dbw replied:
`"abc""def"` will be parsed into `abc"def` in the original code.

SytzeAndr replied:
ah, I was assuming otherwise. Good that you mention this. Apparently this is not covered in the tests, since the build passes.

SytzeAndr replied:
so to be correct: whenever we encounter `""` we should interpret this as a quote character, and when we encounter `"` we should interpret it as an escape quote

daxian-dbw commented:
I believe adopting and reusing the `StringBuilder` is the right thing to do, but I don't think the `StringReader` will do you any benefit in perf because you are essentially replacing array indexing (`csv[i]`) with more expensive method calls (`reader.Peek` and `reader.Read`).

How about just refactoring on top of the existing code instead of writing new code? For example, replace the `switch` statement with `if/else`; restructure the `case '"'` block to reduce the nested blocks; adopting `StringBuilder` like you already do. The existing code can be changed to something like the following, then we can add comments to make it more readable.
```
            bool inQuotes = false;
            for (int i = 0; i < csv.Length; i++)
            {
                char ch = csv[i];
                if (ch == Delimiter)
                {
                    if (inQuotes)
                    {
                        wordBuffer.Append(ch);
                    }
                    else
                    {
                        result.Add(wordBuffer.ToString());
                        wordBuffer.Clear();
                    }
                }
                else if (ch == '"')
                {
                    if (!inQuotes)
                    {
                        // Comments to be added
                        inQuotes = true;
                        continue;
                    }

                    if (i == csv.Length - 1)
                    {
                        // Comments to be added
                        result.Add(wordBuffer.ToString());
                        wordBuffer.Clear();
                        break;
                    }
                    else if (csv[i+1] == '"')
                    {
                        // Comments to be added
                        wordBuffer.Append('"');
                        i++;
                    }
                    else
                    {
                        // Comments maybe?
                        inQuotes = false;
                    }
                }
                else
                {
                    wordBuffer.Append(ch);
                }
            }

            if (wordBuffer.Length > 0)
            {
                result.Add(wordBuffer.ToString());
            }
```

SytzeAndr commented:
@daxian-dbw I like the use of a `reader` instead of array indices, since then the flow becomes less based on values and more on logical statements, which I think is easier to read, easier to extend and results in cleaner code, since it is more restricted and less interpretations are possible. 

I don't think the `peek` and `read` differ that much in performance with respect to direct array index accessing. Readers are made to read lines and characters one-by-one from a String or file and contains some optimized code to do this efficiently. I'm not sure about the reader implementation, but I can believe `read` will be compiled to something that corresponds with a simple index lookup. 

Then again, as @iSazonov commented, the performance should not really be an issue, 

> This code is very rarely used and only for importing a few aliases. 

So if it makes a tiny performance boost with the cost of readability, I would prefer the better readable version, 

As far as I can see, the method complexity per character read (the different paths per character read which can occur) of your implementation is at 7 (2+4+1 paths). The reader implementation has a complexity of 4 (1 + 2 + 1). 

I prefer the reader implementation, but since this is quite opinion based, if the community prefers the `for` loop I'm fine with implementing a `for` loop based implementation.

daxian-dbw commented:
> your implementation is at 7 (2+4+1 paths). The reader implementation has a complexity of 4 (1 + 2 + 1).

You current implementation doesn't handle `"abc""def"` correctly. This should be parsed to `abc"def`, but your implementation will parse it to `abcdef`.
You will need to add 1 or 2 more `if` checks within `if (nextChar == '"')` to fix your current implementation.
The sample code I pasted above can be easily changed to use 6 `if` paths altogether. So there is no much difference if you define maintainability as the number of code paths.

IMHO, the switch statement used in the original implementation and the nested `if` blocks due to the switch is the root cause of bad readability/maintainability, not the fact that it's using array indexing.

SytzeAndr committed:
doesn't escape when two quotes are followed, untill -> until

SytzeAndr committed:
(char) before peek() and codefactor bracket warning fix

SytzeAndr committed:
codefactor: new line before comment

SytzeAndr commented:
I've added an implementation which should resolve the issue regarding parsing `"abc""def"`. @daxian-dbw what do you think, should I implement tests?

iSazonov commented:
> should I implement tests?

Yes, we should cover all code by tests.

iSazonov commented:
> I like the use of a reader instead of array indices

I think StringReader complicates the code and doesn't add readability. Using StringReader looks like personal preference.

SytzeAndr commented:
Since the `csv.cs` is only called in `GetAliasesFromFile()` in `importAliasCommand.cs`, I'm thinking of refactoring the method such that we don't need the seperate `csv.cs`. 

It seems `GetAliasesFromFile()` does use readers. This makes sense as a `reader` is able to read a line from a file faster than a `for` loop would be able to. I would say it makes more sense to use a reader as well for the individual line parsing, for the sake of consistency throughout the method. However, according to the discussion on this Pull Requests, using indices are preferred over readers. As I said before, I'm fine with implementing indices instead if that is preferred, but I do want to know that you are aware that we are mixing two ways of reading text which might be more confusing than to use the `reader` for all reading in the method.

Before I start with the actual implementation, I want to know if I'm on the same line as the community by asking if we agree on the following, to prevent I will be working on something that would get rejected anyways.

* I want to get rid of having the file called `csv.cs` given its current functionality. I think it is confusing at is is not handling all the csv parsing within PowerShell but does have a very general name. 
* Creating the csv file can be called statically. There is no need to create a CSVhelper object to parse a line in the current implementation.
* A check to Ignore blank lines and comment lines is performed in both CSVhelper and `GetAliasesFromFile()` and could be merged such that there's a single point in the code where this violation is checked.
* `GetAliasesFromFile()` is too long and complex and can be split up. Functions can be created for the subprocesses (such as creating the options, creating the alias, checking for violation), such that we reduce complexity and it becomes easier to understand what the code does.
* the error handling is all over the place and could be replaced by a try catch to keep it centralized and to keep it easy to read what the different error outcomes are and how these are handled.

As a start, I've spent some time to make some pseudo-ish-code based on the current implementation of what I think is a refactor that makes more sense while keeping the functionality intact. If we agree, I'd like to hear if you want me to open a new one or that I should keep working within this PR since it is all related.

```
GetAliasesFromFile() {
    result = new collection()
    // linenumber is used for error handling
    Int lineNumber = 0;

    using (StreamReader reader = OpenFile(out filePath, isLiteralPath)) {
    	linenumber = 0
   	
        while ((line = reader.ReadLine()) != null) {
        	++lineNumber
        	try {
        		// parse the line, throw exception if invalid
		        parsedLine = parseCSVLine(line)

			// check for any further violations which is able to throw an exception
			
			isValidParsedLine(parsedLine)

			// if valid add to results
	    		result.add(constructAlias(parsedLine))
	    	} 
	    	catch (InvalidLineException e ) {
	    		// for some cases, like blank line or commented line, we should do nothing and just skip the line
		    } 
		    catch (ArgumentException e) {
		    	// ArgumentException can happen when the options object cannot be created
		    	// we should write an error, create no alias and continue
		    } 
		    catch (NoFourValuesException e) {
		    	// if not four values, do ThrowTerminatingError(errorRecord) just like old code
		    	// code below is copy pasted from importAliasCommand.cs
		    	string message = StringUtil.Format(AliasCommandStrings.ImportAliasFileInvalidFormat, filePath, lineNumber);

		        FormatException formatException =
		            new FormatException(message);

		        ErrorRecord errorRecord =
		            new ErrorRecord(
		                formatException,
		                "ImportAliasFileFormatError",
		                ErrorCategory.ReadError,
		                filePath);

		        errorRecord.ErrorDetails = new ErrorDetails(message)
			ThrowTerminatingError(errorRecord)
		    }		       	
        }	
    }
    return result
}


// this method should correspond with the ParseCsv method in Csv.cs
parseLine(line) {
	if (line.Length == 0 || OnlyContainsWhitespace(line) || line[0] == '#') {
		throw new InvalidLineException()
	}
	// do the line parsing as implemented in CSVHelper.cs
	...
	...
	// return a collection representing the csv string that is parsed
	return result
}

constructAlias(parsedLine) throws argException {
	// create options out of parsed line i.e. a collection object. This might throw an exception
	options = (ScopedItemOptions)Enum.Parse(typeof(ScopedItemOptions), values[3], true);

	// add description (which is the second term of parsedLine) if not null or empty, just like old code

	if (!string.IsNullOrEmpty(values[2]))
        {
            newAlias.Description = values[2];
        }

	// create and return the the actual alias object
	return new AliasInfo(parsedLine, options)
}

// note: I could migrate this violation into parseCSVLine, but then the method would become less general and needs to be renamed.
isValidParsedLine(parsedLine) {
	if(parsedLine.count != 4) {
	    throw NoFourValuesException()
	}
}
```

iSazonov commented:
I wonder that you are going to refactor this old and _very rarely_ used code, but it is permissible for   community project and I will be happy to help you.

First we need to make sure that this code is fully covered with tests and add new tests in another PR if needed.
Remember that our basic rules are to keep PRs as small as possible and split style and functional changes.

We could consider using code from CSVCommand.cs to parse csv and then remove the parser from csv.cs at all. I don’t like to just rename the file because it violates history.
Using static and indexers is good I believe.
We should avoid throws if possible.

SytzeAndr commented:
@iSazonov thanks for your comment and nice to hear it is a permissible community project. So I assume the order would then be to first make tests in a seperate PR (since we have found out the existing tests are not sufficient) and then open up this PR, including the refactorization as I described in my earlier comment.

> I wonder that you are going to refactor this old and very rarely used code

I'm curious about the motive of this statement. Since PowerShell embraces backwards compatibility (so regardless of how rarely a feature is used, it should always be in any newer version) we should include old functionality and thus any piece of code can only be replaced with a newer version: there is no discard option. If we consider code to be messy, it will therefore stay messy forever if we don't see the point of cleaning it up. Shouldn't the 'backwards-compatibility' be a motive to refactor any old code if possible, even if it is very rarely used?

> We should avoid throws if possible

The [code guidelines](https://github.com/PowerShell/PowerShell/blob/master/docs/dev-process/coding-guidelines.md) writes the following about throws:

> Avoid gratuitous exceptions as much as possible. Exception handling can be expensive due to cache misses and page faults when accessing the handling code and data. Finding and designing away exception-heavy code can result in a decent performance win. For example, you should stay away from things like using exceptions for control flow.

There are throws in the code already and IMO it makes an easy to read implementation and reduces complexity. I don't think the example will hurt performance in a significant way (besides, it is used rarely) and testing can be done a lot easier. Do you consider the implementation to be "gratuitous" exceptions? Because if so I'll go for a different approach.

> We could consider using code from CSVCommand.cs to parse csv and then remove the parser from csv.cs at all. I don’t like to just rename the file because it violates history.

I also prefer to remove the parser from `csv.cs` and move/combine all functionality within `CSVCommands.cs`. At the moment the only method in `csv.cs` is the string-parser. If we remove this, there is nothing in `csv.cs` whatsoever. I suggest we will then delete `csv.cs` since the name is confusing.

iSazonov commented:
> Do you consider the implementation to be "gratuitous" exceptions? Because if so I'll go for a different approach.

Please try to do not use throw in hot paths.

Also keep PR as small as possible for fast review.

SytzeAndr committed:
added more tests regarding parsing difficult aliases. However, the test for including a quote character fails, as illegal characters are in the path

SytzeAndr committed:
added tests for parsing invalid files: should throw an error

SytzeAndr commented:
I've added tests in a seperate PR #9247. @iSazonov would you mind taking a look?

SytzeAndr committed:
iSazonov suggestions for testing: added errorID, asserts are now checked by get-alias instead of executing the alias, which is a cleaner method to check whether the alias is imported correctly

SytzeAndr committed:
iSasonov suggestions: adding explicit parameters, fixed type regarding --BeExactly

SytzeAndr committed:
wrong indent fixed

SytzeAndr committed:
removed the initializations and fixed the test for get-alias when none is present. The tests should now pass

SytzeAndr commented:
I've created new pseudocode. I've stepped away from exceptions, except for the `ImportAliasFileFormatError` since that one originates from the old implementation.

```
GetAliasesFromFile() {
    result = new collection()
    // linenumber is used for error handling
    Int lineNumber = 0;

    using (StreamReader reader = OpenFile(out filePath, isLiteralPath)) {
        linenumber = 0
       
        while ((line = reader.ReadLine()) != null) {
            ++lineNumber

            // skip line if it meets some conditions
            if(lineShouldBeSkipped(line)) {
                continue
            }
            
            parsedLine = parseLine(line)

            // if result is valid add to results
            if(isValidParsedLine(parsedLine)) {            
                result.add(constructAlias(parsedLine))
            }
        }    
    }
    return result
}


// this method should correspond with the ParseCsv method in Csv.cs
parseLine(line) {
    // do the line parsing as implemented in CSVHelper.cs
    ...
    ...
    // return a collection representing the csv string that is parsed
    return result
}

constructAlias(parsedLine) throws argException {
    // create options out of parsed line i.e. a collection object. This might throw an exception
    options = (ScopedItemOptions)Enum.Parse(typeof(ScopedItemOptions), values[3], true);

    // add description (which is the second term of parsedLine) if not null or empty, just like old code

    if (!string.IsNullOrEmpty(values[2]))
        {
            newAlias.Description = values[2];
        }

    // create and return the the actual alias object
    return new AliasInfo(parsedLine, options)
}

lineShouldBeSkipped(line) {
    //if line is empty or a comment, return true
    return line.Length == 0 || OnlyContainsWhitespace(line) || line[0] == '#'
}

isValidParsedLine(parsedLine) {
    // if options object cannot be created
    if(optiongoeswrong) {
        //write an error, create no alias and continue
        ...
        return false
    }

    if(values != 4) {
        // if not four values, do ThrowTerminatingError(errorRecord) with ImportAliasFileFormatError, just like old implementation
    }

    return true
}
```

SytzeAndr committed:
iSazonov suggestions about comments, and I renamed some variables

SytzeAndr committed:
reference to variable instead of to a string

SytzeAndr committed:
instead of adding the invalid aliases to the content to create an invalid file, it now creates a new file only containing the invalid alias

SytzeAndr committed:
first version of an implementation base
d upon the pseudocode from my last comment in #9091. Options has been done a little bit differently at th
e moment

SytzeAndr committed:
removed the unnecessary extra check

SytzeAndr committed:
filepath now according to old implementation

SytzeAndr committed:
fixed build issue

SytzeAndr committed:
remove commented code

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/ImportAliasCommand.cs:
```diff
@@ -290,104 +291,244 @@ private Collection<AliasInfo> GetAliasesFromFile(bool isLiteralPath)
-                            new FormatException(message);
+            try
+            {
+                options = (ScopedItemOptions)Enum.Parse(typeof(ScopedItemOptions), parsedLine[3], true);
```
Consider using `Enum.TryParse()`.

SteveL-MSFT committed:
Update test/powershell/Modules/Microsoft.PowerShell.Utility/Import-Alias.Tests.ps1

Co-Authored-By: SytzeAndr <s.p.e.andringa@student.tudelft.nl>

SteveL-MSFT committed:
Update test/powershell/Modules/Microsoft.PowerShell.Utility/Import-Alias.Tests.ps1

Co-Authored-By: SytzeAndr <s.p.e.andringa@student.tudelft.nl>

SteveL-MSFT committed:
Update test/powershell/Modules/Microsoft.PowerShell.Utility/Import-Alias.Tests.ps1

Co-Authored-By: SytzeAndr <s.p.e.andringa@student.tudelft.nl>

SteveL-MSFT committed:
Update test/powershell/Modules/Microsoft.PowerShell.Utility/Import-Alias.Tests.ps1

Co-Authored-By: SytzeAndr <s.p.e.andringa@student.tudelft.nl>

SteveL-MSFT committed:
Update test/powershell/Modules/Microsoft.PowerShell.Utility/Import-Alias.Tests.ps1

Co-Authored-By: SytzeAndr <s.p.e.andringa@student.tudelft.nl>

SteveL-MSFT committed:
Update test/powershell/Modules/Microsoft.PowerShell.Utility/Import-Alias.Tests.ps1

Co-Authored-By: SytzeAndr <s.p.e.andringa@student.tudelft.nl>

SteveL-MSFT committed:
Update test/powershell/Modules/Microsoft.PowerShell.Utility/Import-Alias.Tests.ps1

Co-Authored-By: SytzeAndr <s.p.e.andringa@student.tudelft.nl>

SteveL-MSFT committed:
Update test/powershell/Modules/Microsoft.PowerShell.Utility/Import-Alias.Tests.ps1

Co-Authored-By: SytzeAndr <s.p.e.andringa@student.tudelft.nl>

SteveL-MSFT committed:
Update test/powershell/Modules/Microsoft.PowerShell.Utility/Import-Alias.Tests.ps1

Co-Authored-By: SytzeAndr <s.p.e.andringa@student.tudelft.nl>

SteveL-MSFT committed:
Update test/powershell/Modules/Microsoft.PowerShell.Utility/Import-Alias.Tests.ps1

Co-Authored-By: SytzeAndr <s.p.e.andringa@student.tudelft.nl>

SteveL-MSFT committed:
Update test/powershell/Modules/Microsoft.PowerShell.Utility/Import-Alias.Tests.ps1

Co-Authored-By: SytzeAndr <s.p.e.andringa@student.tudelft.nl>

SteveL-MSFT committed:
Update test/powershell/Modules/Microsoft.PowerShell.Utility/Import-Alias.Tests.ps1

Co-Authored-By: SytzeAndr <s.p.e.andringa@student.tudelft.nl>

SteveL-MSFT committed:
Update test/powershell/Modules/Microsoft.PowerShell.Utility/Import-Alias.Tests.ps1

Co-Authored-By: SytzeAndr <s.p.e.andringa@student.tudelft.nl>

SytzeAndr committed:
removed the afterAll, tests are now with parameters

SytzeAndr committed:
resolved merge conflicts: checkoing for null one level higher for the test when the alias is not imported yet

SytzeAndr committed:
iSazonov suggestions, regarding spacing

SytzeAndr committed:
replaced all tabs with four spaces

adityapatwardhan committed:
Update test/powershell/Modules/Microsoft.PowerShell.Utility/Import-Alias.Tests.ps1

Co-Authored-By: SytzeAndr <s.p.e.andringa@student.tudelft.nl>

SytzeAndr committed:
removed the extra bracket introduced

SytzeAndr committed:
The bug was due to checking if the options was None. Apparently options in general parses to None also if the options was created with succes. I changed the check to a boolean one and by returning a <options,bool> pair instead of only an options object.

SytzeAndr committed:
codefactor suggestions

SytzeAndr committed:
more codefactor suggestions

SytzeAndr committed:
codefactor fixes

SytzeAndr committed:
codefactor fixes, add extra empty line

SytzeAndr committed:
removal of csv.cs since its functionality is transfered to ImportAliasCommand.cd

SytzeAndr committed:
removed reader, implementation is a for loop again

SytzeAndr committed:
made some methods static

SytzeAndr committed:
more codefactor regarding spaces and newlines

SytzeAndr committed:
removed spaced codefactor was complaining about

SytzeAndr committed:
removed casts to char

SytzeAndr committed:
for loop bracket on new line

SytzeAndr committed:
moved static methods to the top of the file. Did some renaming regarding capitals for function names

SytzeAndr commented:
Tests are passing (locally at least, still have to wait for CI). 

CodeFactor is complaining about hungarian notation naming (I used `is` and `in` for naming), but since the rules will be changing soon I don't know if that needs to be fixed or not.

CodeQuality wants methods to be static, but some of them mentioned can't be, like `isValidParsedLine` which used the method `ThrowTerminatingError` which can't be called statically. I made some of them static but not all.

iSazonov commented:
@SytzeAndr No need to fix this style issues.

SytzeAndr commented:
Can I get a new review then?

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/ImportAliasCommand.cs:
```diff
@@ -290,103 +378,118 @@ private Collection<AliasInfo> GetAliasesFromFile(bool isLiteralPath)
+                }
 
-                    Collection<string> values = csvHelper.ParseCsv(line);
+                reader.Dispose();
```
We use `using` which will do the dispose.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/ImportAliasCommand.cs:
```diff
@@ -290,103 +378,118 @@ private Collection<AliasInfo> GetAliasesFromFile(bool isLiteralPath)
+            bool succesfullParse = false;
+            try 
+            {
+                succesfullParse = Enum.TryParse<ScopedItemOptions>(parsedLine[3], out options);
```
If we use TryParse we need to remove `try-catch` and use `if` statement.

SytzeAndr replied:
`TryParse<TEnum>(String, TEnum) is identical to the Parse(Type, String) method, except that instead of throwing an exception, it returns false if the conversion fails. It eliminates the need for exception handling when parsing the string representation of an enumeration value.`

I guess that means I'll have to remove the `WriteError` part since that takes an exception and these are not thrown anymore. That will change functionality a bit, is that a problem? 
Is it fine to just skip (so let it return false) instead of writing an error?

SytzeAndr replied:
If I remove the argException it becomes impossible to pass the test regarding `Should throw an error when reading more than four values`. 

As far as I can see: using `if` and skipping `try/catch` is not possible without failing this test since it expects an exception.

Is it okay to skip the `tryParse` and switch it to a normal `Parse`, or do you have a different suggestion?

```
Should throw an error when reading more than four values 41ms
Expected an exception, with FullyQualifiedErrorId 'ImportAliasFileFormatError,Microsoft.PowerShell.Commands.ImportAliasCommand' to be thrown, but no exception was thrown.
```

iSazonov replied:
Our preference is to avoid exceptions if possible. If an API allows this we should use it. Changing error reporting is not a breaking change. I think the test will remain unchanged (using -Error Stop will cause the exception)

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/ImportAliasCommand.cs:
```diff
@@ -290,103 +378,118 @@ private Collection<AliasInfo> GetAliasesFromFile(bool isLiteralPath)
-                    ScopedItemOptions options = ScopedItemOptions.None;
+            if (!string.IsNullOrEmpty(parsedLine[2]))
+            {
+                newAlias.Description = parsedLine[2];
```
It is better use a temporary variable for `parsedLine[2]`.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/ImportAliasCommand.cs:
```diff
@@ -290,103 +378,118 @@ private Collection<AliasInfo> GetAliasesFromFile(bool isLiteralPath)
+        {
+            if (!optionsParsedSuccesfully)
+            {           
+                return false;
```
This place demonstrates that the idea of ​​using `KeyValuePair`  is unjustified.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/ImportAliasCommand.cs:
```diff
@@ -290,103 +378,118 @@ private Collection<AliasInfo> GetAliasesFromFile(bool isLiteralPath)
             Collection<AliasInfo> result = new Collection<AliasInfo>();
 
+            // it seems like filePath is always null and never defined, so we can remove it as far as I could see
+            // bug keeping it for now because tests are failing
```
No, see line 528.

SytzeAndr replied:
outdated comment, should have been removed

SytzeAndr commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/ImportAliasCommand.cs:
```diff
@@ -99,6 +100,93 @@ public SwitchParameter Force
+                        
+                        if (nextChar == '"') 
+                        {
+                            if (i + 1 < csvTrimmed.Length && csvTrimmed[i + 1] == '"')
```
I'm not sure if I understand you correctly, but as far as I can see `i + 1 < csvTrimmed.Length` is checked first, which directly implies that `csvTrimmed[i + 1] == '"'` can never exceed bounds.

iSazonov replied:
Ah, sorry.

SytzeAndr committed:
Merge branch 'master' into csv_refactor

SytzeAndr committed:
iSazonov suggestions

SytzeAndr committed:
changed to else if

SytzeAndr committed:
Merge branch 'csv_import-alias_tests' into csv_refactor

SytzeAndr committed:
tests are failing locally, I reverted some things but not sure why they are suddenly failing

SytzeAndr committed:
Merge branch 'master' into csv_refactor

SytzeAndr committed:
failing tests, the problem is that options is checked before the four values somehow. That throws an error instead of skipping the line

SytzeAndr committed:
changed order of error checking, now tests pass

SytzeAndr committed:
got rid of the try catch

SytzeAndr commented:
Sorry for the delay. I managed to get rid of the `try/catch` and `KeyValuePair` in `CreateItemOptions`. Tests are passing locally. Can I get a new review?

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/ImportAliasCommand.cs:
```diff
@@ -99,6 +100,93 @@ public SwitchParameter Force
+        private static bool LineShouldBeSkipped(string line)
+        {
+            // if line is empty or a comment, return true
+            return line.Length == 0 || OnlyContainsWhitespace(line) || line[0] == '#';
```
```suggestion
            return line.Length == 0 || line[0] == '#' || OnlyContainsWhitespace(line);
```

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/ImportAliasCommand.cs:
```diff
@@ -99,6 +100,93 @@ public SwitchParameter Force
+
+            foreach (char c in line)
+            {
+                if (char.IsWhiteSpace(c) && c != '\n' && c != '\r')
```
Now we use StreamReader.Readline() - it skip EOL chars.

And now we can replace the method with ReadOnlySpan<char>.isWhitespace()
https://source.dot.net/#System.Private.CoreLib/shared/System/MemoryExtensions.cs,28

SytzeAndr replied:
Just to make sure, you want me to use a reader instead of the `for`? Since you previously said

>I think StringReader complicates the code and doesn't add readability. Using StringReader looks like personal preference.

iSazonov replied:
My current comment is only that make no sense to check `c != '\n' && c != '\r'`.
Previous comment was that we could not use StreamReader at all.

SytzeAndr replied:
clear, sorry for the confusion.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/ImportAliasCommand.cs:
```diff
@@ -99,6 +100,93 @@ public SwitchParameter Force
+      
+        private static Collection<string> ParseCsvLine(string csv)
+        {        
+            string csvTrimmed = csv.Trim();
```
We could use ReadOnlySpan<char>.Trim()
https://source.dot.net/#System.Private.CoreLib/shared/System/MemoryExtensions.Trim.cs,730

SytzeAndr replied:
I can't find a method for `ReadOnlySpan<string>.Trim()`, there seems only to be one for `char`. 
[This documentation](https://docs.microsoft.com/en-us/dotnet/api/system.memoryextensions.trim?view=netstandard-2.1) also only shows a method for `char`.

Shall I leave it like this?

vexx32 replied:
ReadOnlySpan<char> is a fairly analogous standin for a straight `string`. A `Span` (ReadOnly or not) is a sequence of data. A sequence of characters... bit like a string ;)

I haven't looked at all the surrounding code, but conceivably you could have most / all of these `string` variables changed to `ReadOnlySpan<char>` with only minor changes, not 100% sure exactly to what extent Ilya is recommending the change here.

iSazonov replied:
```powershell
        private static Collection<string> ParseCsvLine(ReadOnlySpan<char> csv) 
        {         
             ReadOnlySpan<char> csvTrimmed = csv.Trim();
```

After we move to .Net Core 3.0 we could use StreamReader.Readline() overload returned `ReadOnlySpan<char>`.

iSazonov replied:
Now we moved to .Net Core and can use StreamReader.Readline() overload returned ReadOnlySpan<char>.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/ImportAliasCommand.cs:
```diff
@@ -289,104 +377,101 @@ private bool VerifyShadowingExistingCommandsAndWriteError(string aliasName)
         {
             StreamReader result = null;
 
+            // so remove this line right
```
Please remove the comment or make it more useful.

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Import-Alias.Tests.ps1:
```diff
@@ -2,82 +2,117 @@
-		{Export-Alias  $fulltestpath abcd*} | Should -Not -Throw
-		{Import-Alias $fulltestpath  -Force} | Should -Not -Throw
+    It "Import-Alias From Exported Alias File Aliases Already Exist using force should not throw"{
+        {Export-Alias  $fulltestpath abcd*} | Should -Not -Throw
```
```suggestion
        { Export-Alias  $fulltestpath abcd* } | Should -Not -Throw
```

SytzeAndr replied:
this file shouldn't have changed at all in this PR, I'll try to revert it back

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Import-Alias.Tests.ps1:
```diff
@@ -2,82 +2,117 @@
-		{Import-Alias $fulltestpath  -Force} | Should -Not -Throw
+    It "Import-Alias From Exported Alias File Aliases Already Exist using force should not throw"{
+        {Export-Alias  $fulltestpath abcd*} | Should -Not -Throw
+        {Import-Alias $fulltestpath  -Force} | Should -Not -Throw
```
```suggestion
        { Import-Alias $fulltestpath  -Force } | Should -Not -Throw
```

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Import-Alias.Tests.ps1:
```diff
@@ -2,82 +2,117 @@
+    }
+
+    It "Should classify an alias as non existent when it is not imported yet" {
+        {Get-Alias -Name invalid_alias -ErrorAction Stop} | Should -Throw -ErrorId 'ItemNotFoundException,Microsoft.PowerShell.Commands.GetAliasCommand'
```
```suggestion
        { Get-Alias -Name invalid_alias -ErrorAction Stop } | Should -Throw -ErrorId 'ItemNotFoundException,Microsoft.PowerShell.Commands.GetAliasCommand'
```

SytzeAndr committed:
Update src/Microsoft.PowerShell.Commands.Utility/commands/utility/ImportAliasCommand.cs

Co-Authored-By: Ilya <darpa@yandex.ru>

SytzeAndr committed:
small suggestions fix

SytzeAndr committed:
reverted alias tests to master version

SytzeAndr committed:
typo in implementation, build broke

SytzeAndr committed:
removed unnecessary checks

anmenaga commented:
@iSazonov @daxian-dbw do you see any pending issues with this PR?

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/ImportAliasCommand.cs:
```diff
@@ -289,104 +377,99 @@ private bool VerifyShadowingExistingCommandsAndWriteError(string aliasName)
+        private ScopedItemOptions CreateItemOptions(Collection<string> parsedLine, string filePath, long lineNumber) 
+        {
+            ScopedItemOptions options;
+            if(!Enum.TryParse<ScopedItemOptions>(parsedLine[3], out options)){
```
```suggestion
            if(!Enum.TryParse<ScopedItemOptions>(parsedLine[3], out options))
            {
```

anmenaga reviewed (commented):
@SytzeAndr Please fix Codacy issues in the new code: specifically about converting 'CreateItemOptions','ConstructAlias','IsValidParsedLine' to be static methods. Thank you.

SytzeAndr committed:
Update src/Microsoft.PowerShell.Commands.Utility/commands/utility/ImportAliasCommand.cs

Co-Authored-By: Ilya <darpa@yandex.ru>

SytzeAndr committed:
ReadOnlySpan suggestion

SytzeAndr committed:
made methods static

SytzeAndr committed:
made methods non static

SytzeAndr commented:
Hi, sorry for the radio silence. I fixed the `ReadOnlySpan` trimming

> @SytzeAndr Please fix Codacy issues in the new code: specifically about converting 'CreateItemOptions','ConstructAlias','IsValidParsedLine' to be static methods. Thank you.

Regarding static methods, I didn't made them static due to the following reasons. I'm not sure why codefactor doesn't pick these up. Let me know if you think they are valid excuses.

* I couldn't change `CreateItemOptions` to static due to a call to a public method `WriteError`. 
* `ConstructAlias` uses `Context` at line 434, which is an internal of the abstract class `InternalCommand.Context` and thus can be derived by calling `this.Context`. I changed this and now codefactor doesn't complain anymore.
* `IsValidParsedLine` can't be made static due to a call to the public void `ThrowTerminatingError`.

SytzeAndr committed:
space after if, codefactor issue

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/ImportAliasCommand.cs:
```diff
@@ -99,6 +100,93 @@ public SwitchParameter Force
+                    while (i < csvTrimmed.Length && inQuotes) 
+                    {
+                        i++;
+                        nextChar = csvTrimmed[i];
```
Boundary after i++ can be = csvTrimmed.Length:
```c#
 while (++i < csvTrimmed.Length && inQuotes)
```

SytzeAndr replied:
Wouldn't that change cause `i` to increment regardless whether the while loop is entered?

SytzeAndr replied:
Wouldn't that introduce a bug? Replacing `i` with `++i` would increment `i` regardless whether `inQuotes` is true or false. In other words, when `i` is at the index of the second `"`, it will increment `i` one additional time, skipping the first character after we escape the quote sequence.

For example, `a"b"cd` would then be parsed to `abd`, which is incorrect as it should be parsed to `abcd`.

anmenaga replied:
@iSazonov , can you please looks at this again? thank you.

iSazonov replied:
My comment is only about that there can be boundary overflow because of `i++` without follow check.

SytzeAndr replied:
That would require some different solution, which I think is solved by replacing line 147 with `i + 1 < csvTrimmed.Length && inQuotes` and line 154 with `(i < csvTrimmed.Length && csvTrimmed[i + 1] == '"')`

anmenaga commented:
@daxian-dbw Can you please review this again? thank you.

SytzeAndr committed:
i -> i + 1 in while check

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/ImportAliasCommand.cs:
```diff
@@ -99,6 +100,93 @@ public SwitchParameter Force
+      
+        private static Collection<string> ParseCsvLine(string csv)
+        {        
+            ReadOnlySpan<char> csvTrimmed = csv.Trim();
```
csv should be ReadOnlySpan<char> to avoid allocation here.

iSazonov commented:
Since there is no activity for a long time and this is too edge case, we can close the PR.
